### DISCUSSION TITLE
MOSAIKS On Microsoft Azure Machine Learning Studio

### DISCUSSION THREAD

**Discussion by CarloBroderick** [microsoft/PlanetaryComputer#211] [Q&A]
Hello MPC Community,

First time, long time. 

My name is Carlo Broderick, I am a grad student at UC Santa Barbara following up on a project done at the Bren School last year in which Cullen Molitor and three other grad students implemented the MOSAIKS process to predict Zambian distinct level maize yield. 

The MOSAIKS process is helpfully outlined in one of MPCs [tutorials](https://github.com/microsoft/PlanetaryComputerExamples/blob/main/tutorials/mosaiks.ipynb). Our team originally attempted to use MPC for our MOSAIKS process; however, we found that it was difficult to reliably access MPC's PyTorch instance and extract features, likely because our process was more computationally intensive than last years. 

After migrating over to Microsoft Machine Learning studio and setting up a VM in Western Europe my team is running into some difficulty implementing the MOSAIKS process on Azure. Namely, our process seems to error intermittently with a message:

"Service unavailable
Our services aren't available right now
We're working to restore all services as soon as possible. Please check back soon."
 
The nature of this error and the intermittency make me wonder if this is a symptom of our requests being throttled down and we are requesting too many STAC items. Would authentication help this error? Would I just need to submit my token along with the HTML requests in the STAC api?

Happy to give more information about the problem if that will help resolution. 

Thanks for your support. 

Love what MPC is all about. 

Best,
CB

**Comment by TomAugspurger** [comment]
> Namely, our process seems to error intermittently with a message:

Can you share

1. What endpoint / service is returning that message?
2. What kind of requests you're making when you get that message?
3. The HTTP status code of the response

> symptom of our requests being throttled down and we are requesting too many STAC items. Would authentication help this error?

We do have some rate limits on the STAC API to keep the service health, but those wouldn't result in that error message. And authentication wouldn't help your rate limit on the STAC API (just the SAS API when you're signing items).

**Reply by CarloBroderick** [reply]
Hello Tom,

Forgive me for the long response time.

Here is a gist of what I am experiencing. 
<script src="https://gist.github.com/CarloBroderick/4e31022048523a2a658f63685bc5a2fc.js"></script>

[link](https://gist.github.com/CarloBroderick/4e31022048523a2a658f63685bc5a2fc)



As to your questions:

Can you share

    What endpoint / service is returning that message?
    What kind of requests you're making when you get that message?
    The HTTP status code of the response




Unfortunately I am new to this whole branch of computer science and API's in general. I will likely be using MPC going forward and I would like to improve the way I interact with the service. Could you explain how I would go about finding answers to these questions?

The API error itself:
`APIError: <!DOCTYPE html PUBLIC '-//W3C//DTD XHTML 1.0 Transitional//EN' 'http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd'><html xmlns='http://www.w3.org/1999/xhtml'><head><meta content='text/html; charset=utf-8' http-equiv='content-type'/><style type='text/css'>body {font-family:Arial; margin-left:40px; }img  { border:0 none; }#content { margin-left: auto; margin-right: auto }#message h2 { font-size: 20px; font-weight: normal; color: #000000; margin: 34px 0px 0px 0px }#message p  { font-size: 13px; color: #000000; margin: 7px 0px 0px0px}#errorref { font-size: 11px; color: #737373; margin-top: 41px }</style><title>Service unavailable</title></head><body><div id='content'><div id='message'><h2>Our services aren't available right now</h2><p>We're working to restore all services as soon as possible. Please check back soon.</p></div><div id='errorref'><span>0HApVZAAAAACfktQNGyA/Q6dWsj0wMJTKRlJBMjMxMDUwNDE3MDUzADkyN2FiZmE2LTE5ZjYtNGFmMS1hMDlkLWM5NTlkOWExZTY0NA==</span></div></div></body></html>`

**Reply by TomAugspurger** [reply]
Thanks. That error message indicates some kind of networking / request failure. You might want to try setting up retries on your STAC requests. See https://github.com/microsoft/planetary-computer-apis/issues/176 and linked issues.
